### PR TITLE
Trade Improvements

### PR DESCRIPTION
**Pull request by Athesdrake**
I rewrote most of the trade system.

TradeState
======
The Trade class have now a state attribute that summarize those removed attributes :
 - `on_invite`
 - `accepted`
 - `alive`
 - `canceled`

The equivalent flags are now:
 - `TradeState.ON_INVITE` an invitation has been received from/sent to the other party.
 - `TradeState.ACCEPTING` the client accepted the invitation and is waiting for the other party to be ready.
 - `TradeState.TRADING` the trade has started.
 - `TradeState.CANCELLED` the trade has been cancelled by one of the parties
 - `TradeState.SUCCESS` the trade finished successfully.

TradeContainer
=========
Added the TradeContainer class which represent the content of the trade. Each party has it's own container: `Trade.exports` for the client's items and `Trade.imports` for the other player's items.

The TradeContainer can have up to 5 slots (trade limit). You can get the amount of item a container has with `TradeContainer.get` (return 0 if the item is not found).
You can get the item on a specific slot using `TradeContainer.getSlot`.
`TradeContainer.add` is not intended to be used directly by the client. Use `Trade.addItem` instead.

Trade.addItem & Trade.removeItem
======================
No more headache , `addItem` and `removeItem` now adds/removes a **quantity** of items and not 1 or 10 of them.

Events
====
The events `on_trade_start`, `on_trade_item_change` and `on_trade_lock` no longer have the trade parameter since it's referencing the current trade.
`on_trade_error` has now the error message in addition to the error code.
The trades are nor anymore stored in the `Player` class but in `Client.trades` for the trade having the trader's pid or in `Client.pending_trades` otherwise.
It helps a lot resolving the the trade in the events.

Bugs
===
Fixed smalls bugs:
 - Player.isGuest
 - advance_bot.py error
 - Packet export cache

### CODE REVIEW

File: advance_bot.py [hunk 1]
```diff
@@ -78,9 +78,9 @@ def run(self, block=True):
 			name = msg.author if len(args)<1 else args[0]
 			profile = await self.getProfile(name)
 			if profile is None:
-				await self.reply("The player doesn't exists or is'nt connected.")
+				await msg.reply("The player doesn't exists or is'nt connected.")
```
Comment (Tocutoeltuco) [inline advance_bot.py]: is'nt should be isn't

File: aiotfm/client.py [hunk 2]
```diff
@@ -56,7 +56,8 @@ def __init__(self, community=0, loop=None):
 
 		self.room = None
 		self.trade = None
-		self.trades = []
+		self.trades = {}
+		self.pending_trades = [] # Trades without pid
```
Comment (Tocutoeltuco) [inline aiotfm/client.py]: This list has no .append call.

File: aiotfm/client.py [hunk 3]
```diff
@@ -207,86 +208,83 @@ class Bot(aiotfm.Client):
...
+			name = packet.readUTF().lower()
 			error = packet.read8()
 
-			if name == "":
-				if self.trade._other.username == name:
-					self.trade._close()
-					self.dispatch('trade_error', self.trade, error)
-					self.dispatch('trade_close', self.trade)
-
+			messages = [
```
Comment (Tocutoeltuco) [inline aiotfm/client.py]: I guess this list would be set in enum

File: aiotfm/client.py [hunk 4]
```diff
@@ -207,86 +208,83 @@ class Bot(aiotfm.Client):
...
+				who = self if index == 0 else self.trade.trader
+
+			self.dispatch('trade_lock', who, locked)
 
 		elif CCC==(31, 10): # Trade complete
 			trade = self.trade
-			self.trade._close()
-			self.dispatch('trade_complete', trade)
+			self.trade = False
+			self.trade._close(success=True)
```
Comment (Tocutoeltuco) [inline aiotfm/client.py]: This will throw AttributeError

File: aiotfm/inventory.py [hunk 5]
```diff
@@ -139,145 +147,184 @@ def cmp(a, b):
...
+		return self.pid == other.pid
+
+	@property
+	def closed(self):
+		"""Returns True if the trade is closed."""
+		return self.state in (TradeState.SUCCESS, TradeState.CANCELLED)
+
+	def _start(self, pid):
+		self.state = TradeState.TRADING
+		self.client.trade = self
```
Comment (Tocutoeltuco) [inline aiotfm/inventory.py]: I guess this line will fit better in between the lines 258 and 259 of aiotfm/client.py.

File: aiotfm/utils/enum.py [hunk 6]
```diff
@@ -89,4 +89,12 @@ class commu(Enum):
 	def __missing__(self, key):
 		if isinstance(key, int):
 			return 'int'
-		return 0
\ No newline at end of file
+		return 0
+
+
+class TradeState(Enum):
```
Comment (Tocutoeltuco) [inline aiotfm/utils/enum.py]: A sixth state would be appreciated, which is basically ERROR (if a player leaves the room, disconnects, turns into shaman...). With the ones there it is enough anyway.

File: advance_bot.py (hunk 1 above)
Comment (Athesdrake) [reply advance_bot.py]: 👀 fixed

File: aiotfm/client.py (hunk 2 above)
Comment (Athesdrake) [reply aiotfm/client.py]: I removed `Client.pending_trades` and it throws an error if the player cannot be resolved into a pid.

File: aiotfm/client.py (hunk 3 above)
Comment (Athesdrake) [reply aiotfm/client.py]: I won't keep them as strings then

File: aiotfm/client.py (hunk 4 above)
Comment (Athesdrake) [reply aiotfm/client.py]: Fixed

File: aiotfm/utils/enum.py (hunk 6 above)
Comment (Athesdrake) [reply aiotfm/utils/enum.py]: I find it useless. When a trade has an error, the event `on_trade_error` is dispatched and the trade is deleted.

Comment (Tocutoeltuco) [reply aiotfm/utils/enum.py]: That's true

File: aiotfm/client.py [hunk 7]
```diff
@@ -482,6 +455,18 @@ def get_channel(self, name):
 			if channel.name==name:
 				return channel
 
+	def get_trade(self, player):
+		if not isinstance(player, (str, Player)):
+			raise TypeError(f"Excepted Player or str types got {type(player)}")
```
Comment (Tocutoeltuco) [inline aiotfm/client.py]: typo

**Pull request merged**